DirectProgramming/DPC++/ParallelPatterns/dpc_reduce/src/main.cpp

```diff
@@ -163,32 +163,36 @@ struct slice_area {
   slice_area(int num_steps) { num = num_steps; }
 
   template <typename T>
-  float operator()(T&& i) {
+  float operator()(T&& i) const {
     float x = ((float)i - 0.5) / (float)num;
     return 4.0f / (1.0f + (x * x));
   };
 };
 
 
 // a way to get value_type from both accessors and USM that is needed for transform_init
-template <typename Unknown>
-struct accessor_traits
+template <typename _Unknown>
+struct accessor_traits_impl
 {
 };
 
-template <typename T, int Dim, sycl::access::mode AccMode, sycl::access::target AccTarget,
-          sycl::access::placeholder Placeholder>
-struct accessor_traits<sycl::accessor<T, Dim, AccMode, AccTarget, Placeholder>>
+template <typename _T, int _Dim, sycl::access::mode _AccMode, sycl::access::target _AccTarget,
+          sycl::access::placeholder _Placeholder>
+struct accessor_traits_impl<sycl::accessor<_T, _Dim, _AccMode, _AccTarget, _Placeholder>>
 {
-    using value_type = typename sycl::accessor<T, Dim, AccMode, AccTarget, Placeholder>::value_type;
+    using value_type = typename sycl::accessor<_T, _Dim, _AccMode, _AccTarget, _Placeholder>::value_type;
 };
 
-template <typename RawArrayValueType>
-struct accessor_traits<RawArrayValueType*>
+template <typename _RawArrayValueType>
+struct accessor_traits_impl<_RawArrayValueType*>
 {
-    using value_type = RawArrayValueType;
+    using value_type = _RawArrayValueType;
 };
 
+template <typename _Unknown>
+using accessor_traits = accessor_traits_impl<typename std::decay<_Unknown>::type>;
```

Comment:
Feel free to get rid of the underscore before template arguments (here and above) if it looks ugly

Reply:
will fix